// @hop-protocol/sdk/dist/src/config/config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.bondableChains = exports.config = exports.metadata = void 0;
const { metadata } = require('./metadata');
exports.metadata = metadata;
const mainnet = require('./mainnet');
const staging = require('./staging');
const kovan = require('./kovan');
const goerli = require('./goerli');
const addresses = {
    mainnet: mainnet.addresses,
    staging: staging.addresses,
    kovan: kovan.addresses,
    goerli: goerli.addresses
};
const chains = {
    mainnet: mainnet.chains,
    staging: staging.chains,
    kovan: kovan.chains,
    goerli: goerli.chains
};
const bonders = {
    mainnet: mainnet.bonders,
    staging: staging.bonders,
    kovan: kovan.bonders,
    goerli: goerli.bonders
};
const fees = {
    USDC: {
        ethereum: 18,
        polygon: 18,
        xdai: 25,
        optimism: 18,
        arbitrum: 18
    },
    USDT: {
        ethereum: 25,
        polygon: 25,
        xdai: 30,
        optimism: 25,
        arbitrum: 25
    },
    DAI: {
        ethereum: 25,
        polygon: 25,
        xdai: 30,
        optimism: 25,
        arbitrum: 25
    },
    MATIC: {
        ethereum: 25,
        polygon: 25,
        xdai: 30,
        optimism: 0,
        arbitrum: 0
    },
    ETH: {
        ethereum: 8,
        polygon: 9,
        xdai: 18,
        optimism: 9,
        arbitrum: 9
    },
    WBTC: {
        ethereum: 25,
        polygon: 25,
        xdai: 30,
        optimism: 25,
        arbitrum: 25
    }
};
const config = {
    addresses,
    chains,
    bonders,
    fees
};
exports.config = config;
exports.bondableChains = ['optimism', 'arbitrum'];
