// @hop-protocol/sdk/dist/src/config/config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.bondableChains = exports.config = exports.metadata = void 0;
const { metadata } = require('./metadata');
exports.metadata = metadata;
const mainnet = require('./mainnet');
const staging = require('./staging');
const kovan = require('./kovan');
const goerli = require('./goerli');
const addresses = {
    mainnet: mainnet.addresses,
    staging: staging.addresses,
    kovan: kovan.addresses,
    goerli: goerli.addresses
};
const chains = {
    mainnet: mainnet.chains,
    staging: staging.chains,
    kovan: kovan.chains,
    goerli: goerli.chains
};
const bonders = {
    mainnet: mainnet.bonders,
    staging: staging.bonders,
    kovan: kovan.bonders,
    goerli: goerli.bonders
};
const fees = {
    USDC: {
        L2ToL1: 9,
        L2ToL2: 9,
        anyToxDai: 25
    },
    USDT: {
        L2ToL1: 18,
        L2ToL2: 18,
        anyToxDai: 18
    },
    DAI: {
        L2ToL1: 18,
        L2ToL2: 18,
        anyToxDai: 18
    },
    MATIC: {
        L2ToL1: 18,
        L2ToL2: 18,
        anyToxDai: 18
    },
    ETH: {
        L2ToL1: 4,
        L2ToL2: 5,
        anyToxDai: 18
    }
};
const config = {
    addresses,
    chains,
    bonders,
    fees
};
exports.config = config;
exports.bondableChains = ['optimism', 'arbitrum'];
